File: LogiOS/TrustarApp/Services/Api/ChatApi.swift
//
//  ChatApi.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/23.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import Combine

struct ChatApi {
    func getData(apiKey: String) -> AnyPublisher<[TalkRoom], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getData, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET)
    }

    func getTalkRooms(apiKey: String) -> AnyPublisher<[TalkRoom], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getTalkRooms, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET)
    }
    
    func getHistories(apiKey: String) -> AnyPublisher<[TalkUser], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getHistories, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET)
    }
    
    func getAllUsers(includeInvalid: Bool, apiKey: String) -> AnyPublisher<[TalkUser], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getAllUsers, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET, [("includeInvalid", includeInvalid)])
    }
    
    func addRoom(body: RoomCreate, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.addRoom, [.ApiKeyAndTokenHeader], body.result(), apiKey, .POST)
    }
    
    func addMember(body: RoomAddMember, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.addMember, [.ApiKeyAndTokenHeader], body.result(), apiKey, .POST)
    }
    
    func sendMessage(body: ChatMessagePending, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.sendMessage, [.ApiKeyAndTokenHeader], msg, apiKey, .POST, [], body.jsonBody())
    }
    
    func getMessagePrevious(roomId: String, messageId: String, pageSize: Int, apiKey: String) -> AnyPublisher<[MessageItem], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getMessagePrevious, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET, [("talkRoomId", roomId), ("messageId", messageId), ("pageSize", pageSize)])
    }
    
    func getMessageNext(roomId: String, messageId: String, pageSize: Int, apiKey: String) -> AnyPublisher<[MessageItem], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getMessageNext, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET, [("talkRoomId", roomId), ("messageId", messageId), ("pageSize", pageSize)])
    }
    
    func getMessageItems(roomId: String, pageSize: Int, apiKey: String) -> AnyPublisher<[MessageItem], NetworkError> {
        let msg: ChatMessage? = nil
        return API.shared.callChatAPI(.getMessageItems, [.ApiKeyAndTokenHeader, .CompanyCdQuery], msg, apiKey, .GET, [("talkRoomId", roomId), ("pageSize", pageSize)])
    }
    
    func updateUserSetting(body: ChatUserSettings, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.updateUserSetting, [.ApiKeyAndTokenHeader], body, apiKey, .POST)
    }
    
    func updateRead(body: MessageItemRows, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.updateRead, [.ApiKeyAndTokenHeader], body.result(), apiKey, .POST)
    }
    
    func updateDelete(body: MessageItemRows, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.updateDelete, [.ApiKeyAndTokenHeader], body.result(), apiKey, .POST)
    }
    
    func registerFirebaseToken(body: DefaultPostContent.Token, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.registerFirebaseToken, [.ApiKeyAndTokenHeader, .CompanyCdQuery], body, apiKey, .POST)
    }
    
    func deleteMessage(body: DefaultPostContent.Message, apiKey: String) -> AnyPublisher<StateResult, NetworkError> {
        return API.shared.callChatAPI(.deleteMessage, [.ApiKeyAndTokenHeader], body, apiKey, .POST)
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/rest/api/ChatApi.kt
package jp.co.toukei.log.trustar.rest.api

import io.reactivex.rxjava3.core.Single
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessagePending
import jp.co.toukei.log.trustar.rest.annotation.InjectHttpApi
import jp.co.toukei.log.trustar.rest.annotation.Type
import jp.co.toukei.log.trustar.rest.model.MessageItem
import jp.co.toukei.log.trustar.rest.model.TalkRoom
import jp.co.toukei.log.trustar.rest.model.TalkUser
import jp.co.toukei.log.trustar.rest.post.ChatUserSettings
import jp.co.toukei.log.trustar.rest.post.DefaultPostContent
import jp.co.toukei.log.trustar.rest.post.MessageItemRows
import jp.co.toukei.log.trustar.rest.post.RoomAddMember
import jp.co.toukei.log.trustar.rest.post.RoomCreate
import jp.co.toukei.log.trustar.rest.response.StateResult
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

interface ChatApi {

    @GET("Talk/GetData?isSelectAll=false")
    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    fun getData(): Single<List<TalkRoom>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Talk/GetTalkRooms")
    fun getTalkRooms(): Single<List<TalkRoom>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Talk/GetHistories")
    fun getHistories(): Single<List<TalkUser>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Account/GetAllUsers")
    fun getAllUsers(@Query("includeInvalid") includeInvalid: Boolean): Single<List<TalkUser>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/AddTalkRoom")
    fun addRoom(@Body body: RoomCreate): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/AddMember")
    fun addMember(@Body body: RoomAddMember): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/SendMessage")
    fun sendMessage(@Body body: ChatMessagePending): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Talk/GetPrevious")
    fun getMessagePrevious(
        @Query("talkRoomId") roomId: String,
        @Query("messageId") messageId: String,
        @Query("pageSize") pageSize: Int,
    ): Single<List<MessageItem>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Talk/GetNext")
    fun getMessageNext(
        @Query("talkRoomId") roomId: String,
        @Query("messageId") messageId: String,
        @Query("pageSize") pageSize: Int,
    ): Single<List<MessageItem>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @GET("Talk/GetMessageItems")
    fun getMessageItems(
        @Query("talkRoomId") roomId: String,
        @Query("pageSize") pageSize: Int,
    ): Single<List<MessageItem>>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/UpdateUserSetting")
    fun updateUserSetting(@Body body: ChatUserSettings): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/UpdateIsRead")
    fun updateRead(@Body body: MessageItemRows): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/UpdateIsDeleted")
    fun updateDelete(@Body body: MessageItemRows): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader, Type.CompanyCdQuery)
    @POST("Talk/RegisterFirebaseToken")
    fun registerFirebaseToken(@Body body: DefaultPostContent): Single<StateResult>

    @InjectHttpApi(Type.ApiKeyAndTokenHeader)
    @POST("Talk/DeleteMessage")
    fun deleteMessage(@Body body: DefaultPostContent): Single<StateResult>
}
